// @salesforce/design-system-react/module/components/utilities/label/index.js
/* eslint-disable react/jsx-curly-brace-presence */

/* Copyright (c) 2015-present, salesforce.com, inc. All rights reserved */

/* Licensed under BSD 3-Clause - see LICENSE.txt or git.io/sfdc-license */
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
var propTypes = {
  /*
   * Assistive Text to use instead of a visible label
   */
  assistiveText: PropTypes.object,

  /**
   * Class names to be added to the label
   */
  className: PropTypes.oneOfType([PropTypes.array, PropTypes.object, PropTypes.string]),

  /*
   * Id of the input associated with this label
   */
  htmlFor: PropTypes.string,

  /*
   * Input Label
   */
  label: PropTypes.string,

  /*
   * Applies label styling for a required form element
   */
  required: PropTypes.bool,

  /**
   * Changes markup of label.
   */
  variant: PropTypes.oneOf(['base', 'static'])
};
/*
 * Form label. This returns null if there is no label text (hidden or shown)
 */

var Label = function Label(_ref) {
  var _ref$variant = _ref.variant,
      variant = _ref$variant === void 0 ? 'base' : _ref$variant,
      label = _ref.label,
      assistiveText = _ref.assistiveText,
      className = _ref.className,
      htmlFor = _ref.htmlFor,
      required = _ref.required;
  var labelText = label || assistiveText && assistiveText.label; // One of these is required to pass accessibility tests

  var subRenders = {
    base: /*#__PURE__*/React.createElement("label", {
      className: classNames('slds-form-element__label', {
        'slds-assistive-text': assistiveText && !label
      }, className),
      htmlFor: htmlFor
    }, required && /*#__PURE__*/React.createElement("abbr", {
      className: "slds-required",
      title: "required"
    }, '*'), labelText),
    static: /*#__PURE__*/React.createElement("span", {
      className: classNames('slds-form-element__label', className)
    }, labelText)
  };
  return labelText ? subRenders[variant] : null;
};

Label.displayName = 'Label';
Label.propTypes = propTypes;
export default Label;
