// @salesforce/design-system-react/lib/components/card/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.idSuffixes = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _header = _interopRequireDefault(require("./private/header"));

var _body = _interopRequireDefault(require("./private/body"));

var _footer = _interopRequireDefault(require("./private/footer"));

var _empty = _interopRequireDefault(require("./empty"));

var _constants = require("../../utilities/constants");

var _getAriaProps = _interopRequireDefault(require("../../utilities/get-aria-props"));

var _getDataProps = _interopRequireDefault(require("../../utilities/get-data-props"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var idSuffixes = {
  body: '__body',
  headerActions: '__header-actions',
  heading: '__heading',
  filter: '__filter-input'
};
/**
 * Cards are used to apply a container around a related grouping of information. It has a header, a body, and an optional footer. It often contains a DataTable or Tile (coming soon). Actions associated with selected items or with all items are included within the header actions. Footer often contains pagination. `aria-` and `data-` props can be provided and will be destructured on the root `article` element.
 */

exports.idSuffixes = idSuffixes;

var Card = function Card(_ref) {
  var _ref$heading = _ref.heading,
      heading = _ref$heading === void 0 ? 'Related Items' : _ref$heading,
      bodyClassName = _ref.bodyClassName,
      children = _ref.children,
      className = _ref.className,
      empty = _ref.empty,
      filter = _ref.filter,
      footer = _ref.footer,
      hasNoHeader = _ref.hasNoHeader,
      header = _ref.header,
      headerActions = _ref.headerActions,
      icon = _ref.icon,
      id = _ref.id,
      style = _ref.style,
      rest = _objectWithoutProperties(_ref, ["heading", "bodyClassName", "children", "className", "empty", "filter", "footer", "hasNoHeader", "header", "headerActions", "icon", "id", "style"]);

  var props = _objectSpread({
    heading: heading,
    bodyClassName: bodyClassName,
    children: children,
    className: className,
    empty: empty,
    filter: filter,
    footer: footer,
    hasNoHeader: hasNoHeader,
    header: header,
    headerActions: headerActions,
    icon: icon,
    id: id,
    style: style
  }, rest);

  var ariaProps = (0, _getAriaProps.default)(props);
  var dataProps = (0, _getDataProps.default)(props);
  var bodyId = id ? id + idSuffixes.body : null;
  var filterId = id ? id + idSuffixes.filter : null;
  var headingId = id ? id + idSuffixes.heading : null;
  var headerActionsId = id ? id + idSuffixes.headerActions : null;
  var resolvedEmpty = empty;

  if (resolvedEmpty === true) {
    // Can be overridden by passing in a node to the empty prop
    resolvedEmpty = /*#__PURE__*/_react.default.createElement(_empty.default, {
      id: id,
      heading: heading
    });
  }

  return /*#__PURE__*/_react.default.createElement("article", _extends({
    id: id,
    className: (0, _classnames.default)('slds-card', className),
    style: style
  }, ariaProps, dataProps), !hasNoHeader && /*#__PURE__*/_react.default.createElement(_header.default, {
    header: header,
    headingId: headingId,
    icon: icon,
    filter: filter,
    filterId: filterId,
    heading: heading,
    headerActions: headerActions,
    headerActionsId: headerActionsId
  }), !resolvedEmpty ? /*#__PURE__*/_react.default.createElement(_body.default, {
    id: bodyId,
    className: bodyClassName
  }, children) : /*#__PURE__*/_react.default.createElement(_body.default, {
    id: bodyId,
    className: bodyClassName
  }, resolvedEmpty), footer ? /*#__PURE__*/_react.default.createElement(_footer.default, null, footer) : null);
}; // ### Display Name
// Always use the canonical component name as the React display name.


Card.displayName = _constants.CARD; // ### Prop Types

Card.propTypes = {
  /**
   * CSS classes to be added to the card body (wraps children).
   */
  bodyClassName: _propTypes.default.oneOfType([_propTypes.default.array, _propTypes.default.object, _propTypes.default.string]),

  /**
   * The main section of the card. It often contains a `DataTable` or `Tile`.
   */
  children: _propTypes.default.node,

  /**
   * CSS classes to be added to the card.
   */
  className: _propTypes.default.oneOfType([_propTypes.default.array, _propTypes.default.object, _propTypes.default.string]),

  /**
   * Replaces the body (that is the children) with the specified empty state, this will also remove header actions, the filter, and the icon. If the default empty state is wanted, set to `true`.
   */
  empty: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.node]),

  /**
   * Adds a filter input to the card header.
   */
  filter: _propTypes.default.node,

  /**
   * Footer often contains pagination.
   */
  footer: _propTypes.default.node,

  /**
   * Allows card to have no header, and ignores header related props altogether.
   */
  hasNoHeader: _propTypes.default.bool,

  /**
   * Allows a custom header (the media object with the icon in the first column). `icon`, `heading` and other props are passed into the media object from Card if present. Use `design-system-react/components/media-object` to create your own custom header.
   */
  header: _propTypes.default.node,

  /**
   * The heading is the name of the related item group and should only contain inline elements.
   */
  heading: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.string]),

  /**
   * Actions to perform on selected items or actions that are not specific to one item such as adding an item. If no group actions are needed, then the number of selected items is often present.
   */
  headerActions: _propTypes.default.node,

  /**
   * Icon associated with the items within the `body`.
   */
  icon: _propTypes.default.node,

  /**
   * Set the HTML `id` of the card. This also sets the `id` of the filter and the header actions.
   */
  id: _propTypes.default.string,

  /**
   * Custom styles to be added to the card.
   */
  style: _propTypes.default.object
};
var _default = Card;
exports.default = _default;
